src/batch/src/executor2/generic_exchange.rs

```diff
@@ -128,60 +136,77 @@ impl<CS: 'static + CreateSource> Executor2 for GenericExchangeExecutor2<CS> {
 
 impl<CS: 'static + CreateSource> GenericExchangeExecutor2<CS> {
     #[try_stream(boxed, ok = DataChunk, error = RwError)]
-    async fn do_execute(mut self: Box<Self>) {
-        loop {
-            if self.source_idx >= self.sources.len() {
-                break;
-            }
-            if self.current_source.is_none() {
-                let proto_source = &self.sources[self.source_idx];
-                let source =
-                    CS::create_source(self.env.clone(), proto_source, self.task_id.clone()).await?;
-                self.current_source = Some(source);
+    async fn do_execute(self: Box<Self>) {
+        let mut sources: Vec<Box<dyn ExchangeSource>> = vec![];
+
+        for prost_source in &self.sources {
+            let source =
+                CS::create_source(self.env.clone(), prost_source, self.task_id.clone()).await?;
+            sources.push(source);
+        }
```

Comment:
```suggestion
        let sources = self.sources.map(|s| {
            async move {
                let src: Box<dyn ExchangeSource> = CS::create_source(self.env.clone(), s, self.task_id.clone()).await?;
                src
            }
        })
        .collect();
```

Reply:
This way of writing seems more advanced, but I ran into a bug that I can't fix, I can't throw hello inside the async block
> [{
> 	"resource": "/home/ecs-user/rust/risingwave/src/batch/src/executor2/generic_exchange.rs",
> 	"owner": "rustc",
> 	"code": {
> 		"value": "E0277",
> 		"target": {
> 			"$mid": 1,
> 			"external": "https://doc.rust-lang.org/error-index.html#E0277",
> 			"path": "/error-index.html",
> 			"scheme": "https",
> 			"authority": "doc.rust-lang.org",
> 			"fragment": "E0277"
> 		}
> 	},
> 	"severity": 8,
> 	"message": "the `?` operator can only be used in an async block that returns `Result` or `Option` (or another type that implements `FromResidual`)\nthe trait `FromResidual<std::result::Result<Infallible, RwError>>` is not implemented for `Box<dyn risingwave_rpc_client::ExchangeSource>`",
> 	"source": "rustc",
> 	"startLineNumber": 142,
> 	"startColumn": 118,
> 	"endLineNumber": 142,
> 	"endColumn": 119,
> 	"relatedInformation": [
> 		{
> 			"startLineNumber": 141,
> 			"startColumn": 24,
> 			"endLineNumber": 144,
> 			"endColumn": 14,
> 			"message": "this function should return `Result` or `Option` to accept `?`",
> 			"resource": "/home/ecs-user/rust/risingwave/src/batch/src/executor2/generic_exchange.rs"
> 		}
> 	]
> }]

Reply:
Ok. I've updated my code in the comment. It's because the item type is Result while what `collect` returns not Result.

Reply:
```suggestion
        let sources = self
            .sources
            .into_iter()
            .map(|s| {
                let env = self.env.clone();
                let task_id = self.task_id.clone();
                tokio::spawn(async move {
                    let src: Box<dyn ExchangeSource> = CS::create_source(env, &s, task_id).await?;
                    Ok(src)
                })
            })
            .collect_vec();
        let sources = try_join_all(sources)
            .await?
            .into_iter()
            .map(|s| s)
            .try_collect::<_, Vec<Box<dyn ExchangeSource>>, RwError>()?;
```

Do you mean like this? Parallelizing the `create_source` 🤔